### ISSUE TITLE
[css-color] Should relative `rgb()`/`rgba()` and relative `hsl()`/`hsla()` serialize exactly as declared?

### ISSUE THREAD

**Issue by weinig** [w3c/csswg-drafts#10327]
In CSS Color 5, the [serialization of relative colors](https://drafts.csswg.org/css-color-5/#serial-relative-color) is defined with the following:

> The serialization of the declared value of a relative color function is a string identifying the color function in all-lowercase, followed by "(from ", followed by the serialization of the declared value of the origin color, followed by a single space, followed by a singly-space-separated list of the arguments to the color function, followed by ")".

Does this mean that when serializing `rgb(from ...)`/`rgba(from ...)` that the specific spelling as declared, `rgb` vs `rgba` should be preserved? Or is ok to canonicalize to one or the other?

Similarly, for `hsl(from ...)`/`hsla(from ...)` that the specific spelling as declared, `hsl` vs `hsla` should be preserved? Or is ok to canonicalize to one or the other?

**Comment by kbabbitt** [comment]
The spec goes on to state:
> The serialization of the result of a relative color function depends on whether the keyword [currentColor](https://drafts.csswg.org/css-color-4/#valdef-color-currentcolor) is the [origin color](https://drafts.csswg.org/css-color-5/#origin-color). If so, the result is serialized as the declared value. This allows the correct value to be used on child elements whose [color](https://drafts.csswg.org/css-values-5/#valdef-attr-color) property has a different value. Otherwise, it is the resolved value, which is a [`<color>`](https://drafts.csswg.org/css-color-5/#typedef-color), as defined in [CSS Color 4 §  15. Serializing `<color>` Values](https://drafts.csswg.org/css-color-4/#serializing-color-values).

I understand this to mean:
1. If the origin color is `currentColor`, then serialize the spelling as declared. (We don't know the origin color and thus cannot canonicalize it.)
2. If the origin color is not `currentColor`, then resolve the relative color and serialize canonically as described in Color 4.

There are few tests with examples for how canonicalization happens in case 2: [color-valid.html](https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-valid.html), [color-valid-rgb.html](https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-valid-rgb.html), [color-valid-hsl.html](https://github.com/web-platform-tests/wpt/blob/master/css/css-color/parsing/color-valid-hsl.html). Notably:
- `hsla(...)` canonicalizes to `rgba(...)`, and `hsl(...)` canonicalizes to `rgb(...)`
- `rgb(...)` is used if the clamped alpha is exactly 1, otherwise `rgba(...)` is used

cc @svgeesus to confirm.

**Comment by annevk** [comment]
cc @LeaVerou

**Comment by cdoublev** [comment]
Fwiw, I would prefer to avoid omitting alpha >= 1 and canonicalizing `rgba?()` and `hsla?()` depending on whether it is omitted, but I do not know the "historical reasons" why browsers need to resolve them when serializing them as components of declared values.

**Comment by weinig** [comment]
I don’t think the historical concerns apply to the relative form as there is no existing use to worry about breaking.

**Comment by LeaVerou** [comment]
Yup, I think it's fine to do whatever is reasonable in this case, since the historical concerns don’t apply.

**Comment by weinig** [comment]
My preference here would be to serialize:

 `rgb(from ...)` and `rgba(from ...)` as `rgb(from ...)`
 `hsl(from ...)` and `hsla(from ...)` as `hsl(from ...)`

This loses no useful information, and is minimal, which generally CSS serialization tries to achieve.

It's not a huge deal, implementation wise, to preserve the specific spelling for serialization, so whatever is decided here is fine.

**Comment by svgeesus** [comment]
> My preference here would be to serialize:
> 
> `rgb(from ...)` and `rgba(from ...)` as `rgb(from ...)` `hsl(from ...)` and `hsla(from ...)` as `hsl(from ...)`

I agree this makes the most sense, and would be happy with that resolution.

**Comment by svgeesus** [comment]
proposed resolution: relative rgb()/rgba() and relative hsl()/hsla() serialize as rgb() or hsl() when alpha is unity.

**Comment by kbabbitt** [comment]
Agenda+ to clarify the intent of the WG in the resolution taken previously.

The question originally posed was, by my read, a narrow matter of: *when* a relative color is serialized as `rgb()`/`rgba()` or `hsl()`/`hsla()`, whether the form without the `a` should be used.

Currently, [per spec](https://drafts.csswg.org/css-color-5/#serial-relative-color), the above only happens when a relative color is based on `currentcolor`. For relative colors that are resolvable at parse time, they serialize as their resolved form. And per the same spec section, the resolved form serializes as `color(srgb ...)` since the mixing color space was srgb or hsl. Blink and Gecko have both shipped this behavior.

Some [WPT changes](https://github.com/web-platform-tests/wpt/commit/072e7fd7d7391bb49370e6100610e529d6c9d2b5#diff-3c11dc425e69e4b485a6e9a265d604152cfaa9ba6578de7777a73443badfa682) landed recently that appear to interpret the resolution more broadly: that relative colors declared in `rgb()`/`rgba()`/`hsl()`/`hsla()` *should* serialize in their declared forms rather than their resolved forms. In fact, the WPT changes go one step further and assert that relative colors declared with other functions such as `lab()` or `lch()` should also serialize in their declared forms rather than resolved forms. (See color-valid-relative-color.html in the linked commit.) 

Unfortunately, I did not realize that 'should' implication at the time we took the resolution, nor did I realize that it would constitute a break from shipped behavior. :|

The question of whether to serialize in declared or resolved form is asked more directly in #10305, which the WG has not resolved on. Because that would be a breaking change, it's not clear to me we can make it absent compat analysis. I'm happy to do that analysis; however these changes also impact Interop2024, and the remaining runway for that is short. I can't promise I'd be able to get good enough data for the WG to resolve and implementations to adopt the change by the end of the year.

How does the rest of the WG understand the question and resolution here vs 10305?

cc @emilio

**Comment by weinig** [comment]
@kbabbitt, I think the WPT changes are in line with the changes from https://github.com/w3c/csswg-drafts/issues/10328, where I wrote out a more thorough definition for the serialization of relative colors.

**Comment by kbabbitt** [comment]
@weinig Thanks for the pointer, that issue does cover it, not sure how I missed it. Will take up further discussion there.

**Comment by svgeesus** [comment]
Given that

 - https://github.com/w3c/csswg-drafts/issues/10328

is now closed, with the proposed text accepted by CSS WG resolution, and added to the spec; the examples also added. So I believe that this can be closed, as the current wording reads, for alpha:

> followed by " / " and the alpha component as specified (using the same rules as the color channel arguments) if an alpha component is present, followed by ")".

which seems to be in line with the [resolution here](https://github.com/w3c/csswg-drafts/issues/10327#issuecomment-2162573951).

**Comment by annevk** [comment]
That issue was reopened...

**Comment by svgeesus** [comment]
I know, I re-opened it :) but not because of how to deal with `rgb/rgba` with alpha.
